### Librerias

In [2]:
import numpy as np
import pandas as pd

### Analisis exploratorio

In [3]:
# Cargamos el dataset y mostramos los primeros 10 registros

talks_df = pd.read_csv('raw_data/ted_talks.csv')
talks_df.head(10)
# talks_df.tail(10)

title  \
0      Climate action needs new frontline leadership   
1        The dark history of the overthrow of Hawaii   
2     How play can spark new ideas for your business   
3  Why is China appointing judges to combat clima...   
4     Cement's carbon problem — and 2 ways to fix it   
5  The tragedy of air pollution — and an urgent d...   
6                     The myth of Narcissus and Echo   
7         You deserve the right to repair your stuff   
8  What nature can teach us about sustainable bus...   
9     The origins of blackface and Black stereotypes   

                       author            date   views  likes  \
0        Ozawa Bineshi Albert   December 2021  404000  12000   
1               Sydney Iaukea   February 2022  214000   6400   
2               Martin Reeves  September 2021  412000  12000   
3           James K. Thornton    October 2021  427000  12000   
4             Mahendra Singhi    October 2021    2400     72   
5  Rosamund Adoo-Kissi-Debrah    October 2021  422000  12000   
6            Iseult Gillespie   February 2022  412000  12000   
7            Gay Gordon-Byrne     August 2021  455000  13000   
8                 Erin Meezan   February 2022   66000   1900   
9                  Dwan Reece      March 2019  584000  17000   

                                                link  
0  https://ted.com/talks/ozawa_bineshi_albert_cli...  
1  https://ted.com/talks/sydney_iaukea_the_dark_h...  
2  https://ted.com/talks/martin_reeves_how_play_c...  
3  https://ted.com/talks/james_k_thornton_why_is_...  
4  https://ted.com/talks/mahendra_singhi_cement_s...  
5  https://ted.com/talks/rosamund_adoo_kissi_debr...  
6  https://ted.com/talks/iseult_gillespie_the_myt...  
7  https://ted.com/talks/gay_gordon_byrne_you_des...  
8  https://ted.com/talks/erin_meezan_what_nature_...  
9  https://ted.com/talks/dwan_reece_the_origins_o...

Este dataset contiene registros hasta 2022.

In [4]:
talks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5440 entries, 0 to 5439
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   5440 non-null   object
 1   author  5439 non-null   object
 2   date    5440 non-null   object
 3   views   5440 non-null   int64 
 4   likes   5440 non-null   int64 
 5   link    5440 non-null   object
dtypes: int64(2), object(4)
memory usage: 255.1+ KB


Existen 5440 pero tenemos un dato faltante en la columna de Autores.

In [5]:
# Mostramos el registro del autor que falta
talks_df[talks_df['author'].isnull()]

title author           date  views  likes  \
3039  Year In Ideas 2015    NaN  December 2015    532     15   

                                          link  
3039  https://ted.com/talks/year_in_ideas_2015

Esta charla es un resumen, el cual no contiene un autor y se eliminara para la integridad del dataset.

(Te invito a ver el resumen, es interesante ver como algunas tecnologías evolucionaron rápidamente, pero también algunos problemas se mantuvieron a pesar de los años)


In [6]:
# Dataframe que excluye el registro de autor faltante
talks_clean = talks_df[~talks_df['author'].isnull()]

In [7]:
talks_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5439 entries, 0 to 5439
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   5439 non-null   object
 1   author  5439 non-null   object
 2   date    5439 non-null   object
 3   views   5439 non-null   int64 
 4   likes   5439 non-null   int64 
 5   link    5439 non-null   object
dtypes: int64(2), object(4)
memory usage: 297.4+ KB


Mostramos los datos únicos de cada columna dentro del dataset.

In [8]:
for columna in talks_clean.columns:
    print(f'Valores unicos en la columna {columna}: {talks_clean[columna].nunique()}')

Valores unicos en la columna title: 5439
Valores unicos en la columna author: 4443
Valores unicos en la columna date: 200
Valores unicos en la columna views: 972
Valores unicos en la columna likes: 752
Valores unicos en la columna link: 5439


- Los Valores únicos en la columna author son: 4,443. Lo que indica que existen autores con más de una charla en el dataset.

- Los Valores únicos en la columna title son: 5,439. (Es el número total de registros con los que estamos trabajando)


In [9]:
# Restamos el numero total de registros y el numero de autores unicos para obtener el numero de autores faltantes
talks_clean['author'].count() - talks_clean['author'].nunique() 

996

In [10]:
# Mostrar las celdas que contienen autores repetidos
talks_clean[talks_clean['author'].duplicated()]

# Contamos todos los registros que contienen autores repetidos
# talks_clean[talks_clean['author'].duplicated()].count()

title  \
43                 The tragic myth of the Sun God's son   
58                The myth of Loki's monstrous children   
61                               Human-centered justice   
77    How to deal with your insomnia — and finally g...   
95        How sleep affects what (and how much) you eat   
...                                                 ...   
5435                    The best stats you've ever seen   
5436                        Do schools kill creativity?   
5437                                Greening the ghetto   
5438                                   Simplicity sells   
5439                        Averting the climate crisis   

                         author           date     views    likes  \
43             Iseult Gillespie   January 2022    998000    29000   
58             Iseult Gillespie   January 2022    761000    22000   
61    American Family Insurance   January 2022     11000      342   
77                  Matt Walker  December 2021    311000     9300   
95                  Matt Walker  December 2021    315000     9400   
...                         ...            ...       ...      ...   
5435               Hans Rosling  February 2006  15000000   458000   
5436           Sir Ken Robinson  February 2006  72000000  2100000   
5437              Majora Carter  February 2006   2900000    88000   
5438                David Pogue  February 2006   2000000    60000   
5439                    Al Gore  February 2006   3600000   109000   

                                                   link  
43    https://ted.com/talks/iseult_gillespie_the_tra...  
58    https://ted.com/talks/iseult_gillespie_the_myt...  
61    https://ted.com/talks/american_family_insuranc...  
77    https://ted.com/talks/matt_walker_how_to_deal_...  
95    https://ted.com/talks/matt_walker_how_sleep_af...  
...                                                 ...  
5435  https://ted.com/talks/hans_rosling_the_best_st...  
5436  https://ted.com/talks/sir_ken_robinson_do_scho...  
5437  https://ted.com/talks/majora_carter_greening_t...  
5438  https://ted.com/talks/david_pogue_simplicity_s...  
5439  https://ted.com/talks/al_gore_averting_the_cli...  

[996 rows x 6 columns]

La fecha contiene un formato poco habitual, por lo que se dividirá en mes y año.

In [11]:
talks_clean[['month', 'year']] = talks_clean['date'].str.split(' ', expand=True)

C:\Users\G\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\frame.py:3641: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [12]:
# Eliminar la columna de date
talks_clean.drop(columns=['date'], inplace=True)

C:\Users\G\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [13]:
# Cual es la primer charla en los registros de Ted Talks?

# Ordenamos el dataset por fecha y mostramos el primer registro
talks_clean.sort_values(by=['year'], ascending=True).head(1)

title          author   views  likes  \
738  Love letters to what we hold dear  Debbie Millman  192000   5700   

                                                  link    month  year  
738  https://ted.com/talks/debbie_millman_love_lett...  January  1970

In [28]:
# Numero de platicas por año. De mayor a menor
talks_clean['year'].value_counts(sort=True)

2019    544
2020    501
2017    495
2018    473
2016    399
2021    390
2013    388
2015    376
2014    357
2012    302
2011    271
2010    267
2009    233
2007    113
2008     84
2005     65
2006     49
2003     34
2004     33
2002     26
2022     20
1998      6
2001      5
1970      2
1990      1
1994      1
1991      1
1983      1
1972      1
1984      1
Name: year, dtype: int64

In [15]:
# Cuantos años hay en el dataset?
talks_clean['year'].nunique()

30

In [16]:
# Top 10 ted talks con mas likes
talks_clean.groupby('title')['likes'].sum().sort_values(ascending = False).head(10)


title
Do schools kill creativity?                                            2100000
Your body language may shape who you are                               1900000
Inside the mind of a master procrastinator                             1800000
The power of vulnerability                                             1700000
How great leaders inspire action                                       1700000
How to speak so that people want to listen                             1400000
My philosophy for a happy life                                         1300000
The next outbreak? We're not ready                                     1300000
What makes a good life? Lessons from the longest study on happiness    1200000
Looks aren't everything. Believe me, I'm a model.                      1100000
Name: likes, dtype: int64

In [17]:
# Top 10 ted talks con mas vistas
talks_clean.groupby('title')['views'].sum().sort_values(ascending = False).head(10)


title
Do schools kill creativity?                                            72000000
Your body language may shape who you are                               64000000
Inside the mind of a master procrastinator                             60000000
How great leaders inspire action                                       57000000
The power of vulnerability                                             56000000
How to speak so that people want to listen                             49000000
The next outbreak? We're not ready                                     43000000
My philosophy for a happy life                                         43000000
What makes a good life? Lessons from the longest study on happiness    41000000
Looks aren't everything. Believe me, I'm a model.                      38000000
Name: views, dtype: int64

### Exportar:

Esta linea se dejara comentada para que no se guarde cada vez que se ejecute el notebook

In [18]:
# talks_clean.to_csv('clean_data/ted_talks_clean.csv', index=False)